webhook/resource_url_notifier.go

```diff
@@ -0,0 +1,348 @@
+)
+
+type poster interface {
+	Process(ctx context.Context, queuedAt time.Time, event *livekit.WebhookEvent)
+}
+
+type resourceQueueInfo struct {
+	*resourceQueue
+	key string
+	tqi *utils.TimeoutQueueItem[*resourceQueueInfo]
+}
+
+type ResourceURLNotifierParams struct {
+	HTTPClientParams
+	Logger     logger.Logger
+	Timeout    time.Duration
+	MaxAge     time.Duration
+	MaxDepth   int
+	URL        string
+	APIKey     string
+	APISecret  string
+	FieldsHook func(whi *livekit.WebhookInfo)
+}
+
+// ResourceURLNotifier is a QueuedNotifier that sends a POST request to a Webhook URL.
+// It queues up events per resource (could be egress, ingress, room, participant, track, etc.)
+//   to avoid blocking events of one resource blocking another resource's event(s).
+// It will retry on failure, and will drop events if notification fall too far behind,
+//   either in age or queue depth.
+type ResourceURLNotifier struct {
+	mu            sync.RWMutex
+	params        ResourceURLNotifierParams
+	client        *retryablehttp.Client
+	processedHook func(ctx context.Context, whi *livekit.WebhookInfo)
+
+	resourceQueues            map[string]*resourceQueueInfo
+	resourceQueueTimeoutQueue utils.TimeoutQueue[*resourceQueueInfo]
+	closed                    core.Fuse
+}
+
+func NewResourceURLNotifier(params ResourceURLNotifierParams) *ResourceURLNotifier {
+	if params.Logger == nil {
+		params.Logger = logger.GetLogger()
+	}
+
+	if params.Timeout == 0 {
+		params.Timeout = defaultTimeout
+	}
+	if params.MaxAge == 0 {
+		params.MaxAge = defaultMaxAge
+	}
+	if params.MaxDepth == 0 {
+		params.MaxDepth = defaultMaxDepth
+	}
+
+	rhc := retryablehttp.NewClient()
+	if params.RetryWaitMin > 0 {
+		rhc.RetryWaitMin = params.RetryWaitMin
+	}
+	if params.RetryWaitMax > 0 {
+		rhc.RetryWaitMax = params.RetryWaitMax
+	}
+	if params.MaxRetries > 0 {
+		rhc.RetryMax = params.MaxRetries
+	}
+	if params.ClientTimeout > 0 {
+		rhc.HTTPClient.Timeout = params.ClientTimeout
+	}
+	rhc.Logger = &logAdapter{}
+	r := &ResourceURLNotifier{
+		params:         params,
+		client:         rhc,
+		resourceQueues: make(map[string]*resourceQueueInfo),
+	}
+
+	go r.sweeper()
+	return r
+}
+
+func (r *ResourceURLNotifier) SetKeys(apiKey, apiSecret string) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.params.APIKey = apiKey
+	r.params.APISecret = apiSecret
+}
+
+func (r *ResourceURLNotifier) RegisterProcessedHook(hook func(ctx context.Context, whi *livekit.WebhookInfo)) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.processedHook = hook
+}
+
+func (r *ResourceURLNotifier) getProcessedHook() func(ctx context.Context, whi *livekit.WebhookInfo) {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+	return r.processedHook
+}
+
+func (r *ResourceURLNotifier) QueueNotify(ctx context.Context, event *livekit.WebhookEvent) error {
+	if r.closed.IsBroken() {
+		return errClosed
+	}
+
+	key := eventKey(event)
+
+	r.mu.Lock()
+	var tqi *utils.TimeoutQueueItem[*resourceQueueInfo]
+	rqi := r.resourceQueues[key]
+	if rqi == nil {
```

Comment:
we have to call `Reset` on the timeout queue on every access to bump the deadline. if we do it before releasing this lock the rest of the code can be simpler

we can check that the item hasn't expired like
```go
if rqi == nil || !r.resourceQueueTimeoutQueue.Reset(rqi.tqi) {
```